### PR TITLE
Rethink `AVPlayerItemStatus` handling

### PR DESCRIPTION
**Pull request by BohdanBilonoh**
@aol-public/mobile-sdk-team: Ready for review.

[JIRA Issue](https://jira.ops.aol.com/browse/OMSDK-XXXX)

### CODE REVIEW

File: VideoRenderer/VideoRenderer/SystemPlayerObserver.swift [hunk 1]
```diff
@@ -163,13 +165,13 @@ public final class SystemPlayerObserver: NSObject {
...
             
         case #keyPath(AVPlayerItem.status):
-            let oldStatus = oldValue().flatMap(AVPlayerItemStatus.init)
-            guard let newStatus = newValue().flatMap(AVPlayerItemStatus.init) else {
-                fatalError("Unexpected nil in AVPlayerItem.status value!")
+            guard let newStatus = newValue().flatMap(AVPlayerItemStatus.init) else { fatalError("Unexpected nil in AVPlayerItem.status value!") }
+            switch newStatus {
+            case .unknown: emit(.didChangeItemStatusToUnknown())
+            case .readyToPlay: emit(.didChangeItemStatusToReadyToPlay())
+            case .failed: return emit(.didChangeItemStatusToFailed(error: player.error ?? player.currentItem?.error))
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/SystemPlayerObserver.swift]: Is `error` of the `player` observable?
Maybe we can observe it and instead send failed when we received an error?

**Review by AndriiMoskvin — CHANGES_REQUESTED** [review]

File: VideoRenderer/VideoRenderer/SphereVideoStreamViewController.swift [hunk 2]
```diff
@@ -103,19 +103,15 @@ public class SphereVideoStreamViewController: GLKViewController, RendererProtoco
...
-                                return error
-                            }()
-                            self?.dispatch?(.playbackFailed(error))
-                        default: break
-                        }
+                    case .didChangeItemStatusToFailed(let error):
+                        let error: Error = {
+                            guard let error = error else {
+                                struct UnknownError: Error { let props: Renderer.Props }
+                                return UnknownError(props: props)
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/SphereVideoStreamViewController.swift]: Add a bit more semantic to the name - at least add where this error comes from.

File: VideoRenderer/VideoRenderer/SystemPlayerObserver.swift (hunk 1 above)
Comment (BohdanBilonoh) [reply VideoRenderer/VideoRenderer/SystemPlayerObserver.swift]: In this case we should observe `player` and `currentItem` `error`. Is it ok?

**Review by AndriiMoskvin — CHANGES_REQUESTED** [review]

File: VideoRenderer/VideoRenderer/SystemPlayerObserver.swift (hunk 1 above)
Comment (AndriiMoskvin) [reply VideoRenderer/VideoRenderer/SystemPlayerObserver.swift]: Let's leave only `currentItem.error`

File: VideoRenderer/VideoRenderer/SystemPlayerObserver.swift [hunk 3]
```diff
@@ -8,12 +8,15 @@ public final class SystemPlayerObserver: NSObject {
...
-        case didChangeItemStatus(from: AVPlayerItemStatus?, to: AVPlayerItemStatus)
+        case didChangeItemStatusToUnknown()
+        case didChangeItemStatusToReadyToPlay()
+        case didChangeItemStatusToFailed(error: Error?)
         case didFinishPlayback(withUrl: URL)
         case didChangeLoadedTimeRanges(to: [CMTimeRange])
         case didChangeAverageVideoBitrate(to: Double)
         case didChangeItemDuration(to: CMTime)
         case didChangeAsset(AVAsset)
+        case playerErrored(error: Error)
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/SystemPlayerObserver.swift]: `didReceivedPlayerError`

File: VideoRenderer/VideoRenderer/SystemPlayerObserver.swift [hunk 4]
```diff
@@ -161,15 +169,19 @@ public final class SystemPlayerObserver: NSObject {
...
             
         case #keyPath(AVPlayerItem.status):
-            let oldStatus = oldValue().flatMap(AVPlayerItemStatus.init)
-            guard let newStatus = newValue().flatMap(AVPlayerItemStatus.init) else {
-                fatalError("Unexpected nil in AVPlayerItem.status value!")
+            guard let newStatus = newValue().flatMap(AVPlayerItemStatus.init) else { fatalError("Unexpected nil in AVPlayerItem.status value!") }
+            switch newStatus {
+            case .unknown: emit(.didChangeItemStatusToUnknown())
+            case .readyToPlay: emit(.didChangeItemStatusToReadyToPlay())
+            case .failed: emit(.didChangeItemStatusToFailed(error: player.currentItem?.error))
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/SystemPlayerObserver.swift]: Where is `AVPlayerItem.Error` observing?

**Review by AndriiMoskvin — CHANGES_REQUESTED** [review]

File: VideoRenderer/VideoRenderer/SystemPlayerObserver.swift [hunk 5]
```diff
@@ -8,12 +8,15 @@ public final class SystemPlayerObserver: NSObject {
...
-        case didChangeItemStatus(from: AVPlayerItemStatus?, to: AVPlayerItemStatus)
+        case didChangeItemStatusToUnknown()
+        case didChangeItemStatusToReadyToPlay()
+        case didChangeItemStatusToFailed(error: Error)
         case didFinishPlayback(withUrl: URL)
         case didChangeLoadedTimeRanges(to: [CMTimeRange])
         case didChangeAverageVideoBitrate(to: Double)
         case didChangeItemDuration(to: CMTime)
         case didChangeAsset(AVAsset)
+        case didReceivedPlayerError(error: Error)
```
Comment (AndriiMoskvin) [inline VideoRenderer/VideoRenderer/SystemPlayerObserver.swift]: `didReceivePlayerError(Error)`

**Review by AndriiMoskvin — CHANGES_REQUESTED** [review]

**Review by AndriiMoskvin — APPROVED** [review]

**Pull request merged**